### PR TITLE
GUI 

### PR DESCRIPTION
**Pull request by JiahuiKChen**
GUI for the CodeUSummer2017 chat application, made using Java Swing. 

Comprised of one window featuring mainly a text display of application output, buttons to trigger commands and a text input box below the text display. Each internal panel of the application (root, user, conversation) has its own panel display and the GUI changes panel displays in sync with where the user is in the application.

### CODE REVIEW

File: src/codeu/chat/util/GUI.java [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+        //Text input for user's messages/other conversation commands
+        JTextField convoTextInput = new JTextField(1);
+        convoInputPanel.add(convoTextInput);
+
+        //Command buttons
+        JPanel convoButtonPanel = new JPanel();
+        convoButtonPanel.setLayout(new GridLayout(1, 9));
+        convoInputPanel.add(convoButtonPanel);
+        JButton convoHelp = new JButton("Help");
+        convoHelp.addActionListener(new ActionListener() {
```
Comment (kevmo314) [inline src/codeu/chat/util/GUI.java]: You can probably simplify the code below by adding a method:

```
private JButton createActionButton(String label, Supplier<Void> handleCommand) {
  JButton button = new JButton(label);
  button.addActionListener(new ActionListener() {
    @Override
    public void actionPerformed(ActionEvent e) {
      handleCommand.get();
    }
  });
  return button;
}
```

Then calling

```
convoButtonPanel.add(createActionButton("Help", () -> chat.handleCommand("help")));
```

for all the buttons.

File: src/codeu/chat/util/GUI.java [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+
+        //Text input for user panel commands
+        JTextField userTextInput = new JTextField(1);
+        userInputPanel.add(userTextInput);
+
+        //Command buttons
+        JPanel userButtonPanel = new JPanel();
+        userButtonPanel.setLayout(new GridLayout(2, 7));
+        userInputPanel.add(userButtonPanel);
+        JButton userHelp = new JButton("Help");
```
Comment (kevmo314) [inline src/codeu/chat/util/GUI.java]: These can also be simplified with an additional method (see comment below).

File: src/codeu/chat/util/GUI.java [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+
+        //Text input to sign in/add user
+        JTextField rootTextInput = new JTextField(1);
+        rootInputPanel.add(rootTextInput);
+
+        //Command buttons
+        JPanel rootButtonPanel = new JPanel();
+        rootButtonPanel.setLayout(new GridLayout(1, 6));
+        rootInputPanel.add(rootButtonPanel);
+        JButton rootHelp = new JButton("Help");
```
Comment (kevmo314) [inline src/codeu/chat/util/GUI.java]: These can also be simplified (see below).

File: src/codeu/chat/util/GUI.java (hunk 1 above)
Comment (JiahuiKChen) [reply src/codeu/chat/util/GUI.java]: Code has been simplified. I tried incorporating the output stream changes and panel switching into the `createActionButton` method's ActionListener for the buttons that need to trigger those changes, but it couldn't be done without making a bunch more field variables, so I just left the creation code for those JButtons as is.

**Review by kevmo314 — APPROVED** [review]

**Pull request merged**